java-checks/src/main/java/org/sonar/java/checks/RedundantConstructorsAndMethodsShouldBeAvoidedCheck.java

```diff
@@ -0,0 +1,147 @@
+    List<ReturnStatementTree> returnStatements = method.block().body().stream()
+      .filter(statement -> statement.is(Tree.Kind.RETURN_STATEMENT))
+      .map(ReturnStatementTree.class::cast)
+      .collect(Collectors.toList());
+    boolean onlyReturnsComponents = false;
+    for (ReturnStatementTree returnStatement : returnStatements) {
+      ExpressionTree expression = returnStatement.expression();
+      IdentifierTree identifier;
+      if (expression.is(Tree.Kind.IDENTIFIER)) {
+        identifier = (IdentifierTree) expression;
+      } else if (expression.is(Tree.Kind.MEMBER_SELECT)) {
+        identifier = ((MemberSelectExpressionTree) expression).identifier();
+      } else {
+        continue;
+      }
+      Symbol identifierSymbol = identifier.symbol();
+      boolean identifierMatchesComponent = components.stream().anyMatch(component -> component.symbol().equals(identifierSymbol));
+      onlyReturnsComponents |= identifierMatchesComponent;
+    }
+    return onlyReturnsComponents;
+  }
+
+  public static boolean onlyDoesSimpleAssignments(MethodTree constructor, List<VariableTree> components) {
+    List<VariableTree> parameters = constructor.parameters();
+    if (parameters.size() != components.size()) {
+      return false;
+    }
+    List<StatementTree> statements = constructor.block().body();
+    Set<Symbol> componentsAssignedInConstructor = new HashSet<>();
+    for (StatementTree statement : statements) {
+      Optional<AssignmentExpressionTree> assignment = extractAssignment(statement);
+      if (!assignment.isPresent()) {
+        continue;
+      }
+      ExpressionTree leftHandSide = assignment.get().variable();
+      if (!leftHandSide.is(Tree.Kind.MEMBER_SELECT)) {
+        continue;
+      }
+      Symbol variableSymbol = ((MemberSelectExpressionTree) leftHandSide).identifier().symbol();
+      boolean variableIsAComponent = components.stream().anyMatch(component -> component.symbol().equals(variableSymbol));
+      if (!variableIsAComponent) {
+        continue;
+      }
+      ExpressionTree rightHandSide = assignment.get().expression();
+      if (!rightHandSide.is(Tree.Kind.IDENTIFIER)) {
+        continue;
+      }
+      Symbol valueSymbol = ((IdentifierTree) rightHandSide).symbol();
+      boolean valueIsAParameter = parameters.stream().anyMatch(parameter -> parameter.symbol().equals(valueSymbol));
+      if (valueIsAParameter && variableSymbol.name().equals(valueSymbol.name())) {
+        componentsAssignedInConstructor.add(variableSymbol);
+      }
+    }
+    return componentsAssignedInConstructor.size() == components.size();
+  }
+
+  private static Optional<AssignmentExpressionTree> extractAssignment(StatementTree statement) {
+    if (!statement.is(Tree.Kind.EXPRESSION_STATEMENT)) {
+      return Optional.empty();
+    }
+    ExpressionStatementTree initialStatement = (ExpressionStatementTree) statement;
```

Comment:
directly use a variable `expression`, to not have to reuse `initialStatement.expression()`:

```java
ExpressionTree expression = ((ExpressionStatementTree) statement).expression();
```